File: solutions/0907-koko-eating-bananas/solution.java
class Solution {
   public static int max(int[] arr){
        int max = arr[0];
        for (int i = 0; i < arr.length; i++) if(max<arr[i]) max = arr[i];
        return max;
    }
    public static long timeCalculation(int[] arr, int speed){
        long sum = 0;
        for (int i = 0; i < arr.length; i++) {
            int time = arr[i]/speed;
            if(arr[i]%speed==0) sum+=time;
            else sum = sum+time+1;
        }
        return sum;
    }

    public static int minEatingSpeed(int[] piles, int h) {
        int low = 1;
        int high = max(piles);
        if(h==piles.length) return max(piles);
        else{
            while(low<=high){
                int mid = low+(high-low)/2;
                long time = timeCalculation(piles, mid);
                System.out.println(time);
                if(time>h) low = mid+1;
                else high = mid-1;
            }
        }
        return low;
    }
}

File: solutions/0907-koko-eating-bananas/solution.cpp
class Solution {
public:
    long long calTime(vector<int>& piles, long long cap){
        long long cnt = 0;
        for(long long ele: piles){
            cnt += ele/cap;
            if(ele%cap!=0) cnt++;
        }
        return cnt;
    }
    int minEatingSpeed(vector<int>& piles, long long h) {
        long long l = 1;
        long long maxi = piles[0];
        for(long long ele: piles) maxi = max(maxi, ele);
        long long r = maxi;
        while(l<=r){
            long long mid = l+(r-l)/2;
            long long hrs = calTime(piles, mid);
            if(hrs>h) l = mid+1;
            else r = mid-1;
        }
        return int(l);
    }
};
